substrate/emissions/pallet/src/lib.rs

```diff
@@ -0,0 +1,413 @@
+    StorageMap<_, Identity, NetworkId, bool, ValueQuery>;
+
+  #[pallet::storage]
+  #[pallet::getter(fn last_swap_volume)]
+  pub(crate) type LastSwapVolume<T: Config> = StorageMap<_, Identity, NetworkId, u64, OptionQuery>;
+
+  #[pallet::storage]
+  pub(crate) type GenesisCompleteBlock<T: Config> = StorageValue<_, u64, OptionQuery>;
+
+  #[pallet::genesis_build]
+  impl<T: Config> BuildGenesisConfig for GenesisConfig<T> {
+    fn build(&self) {
+      for (id, stake) in self.networks.clone() {
+        let mut participants = vec![];
+        for p in self.participants.clone() {
+          participants.push((p, stake.0));
+        }
+        Participants::<T>::set(id, Some(participants.try_into().unwrap()));
+        CurrentSession::<T>::set(id, 0);
+        EconomicSecurityReached::<T>::set(id, false);
+      }
+    }
+  }
+
+  #[pallet::hooks]
+  impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+    fn on_initialize(n: BlockNumberFor<T>) -> Weight {
+      if GenesisCompleteBlock::<T>::get().is_none() &&
+        GenesisLiquidity::<T>::genesis_complete().is_some()
+      {
+        GenesisCompleteBlock::<T>::set(Some(n.saturated_into::<u64>()));
+      }
+
+      // we wait 1 extra block after genesis ended to see the changes. We only need this extra
+      // block in dev&test networks where we start the chain with accounts that already has some
+      // staked SRI. So when we check for ec-security pre-genesis we look like we are economically
+      // secure. The reason for this although we only check for it once the genesis is complete(so
+      // if the genesis complete we shouldn't be economically secure because the funds are not
+      // enough) is because ValidatorSets pallet runs before the genesis pallet in runtime.
+      //  So ValidatorSets pallet sees the old state until next block.
+      let gcb = GenesisCompleteBlock::<T>::get();
+      let genesis_ended = gcb.is_some() && (n.saturated_into::<u64>() > gcb.unwrap());
+
+      // we accept we reached economic security once we can mint smallest amount of a network's coin
+      for coin in COINS {
+        let check = genesis_ended && !Self::economic_security_reached(coin.network());
+        if check && <T as CoinsConfig>::AllowMint::is_allowed(&Balance { coin, amount: Amount(1) })
+        {
+          EconomicSecurityReached::<T>::set(coin.network(), true);
+        }
+      }
+
+      // check if we got a new session
+      let mut session_changed = false;
+      let session = ValidatorSets::<T>::session(NetworkId::Serai).unwrap_or(Session(0));
+      if session.0 > Self::session(NetworkId::Serai) {
+        session_changed = true;
+        CurrentSession::<T>::set(NetworkId::Serai, session.0);
+      }
+
+      // update participants per session before the genesis and after the genesis
+      // we update them after reward distribution.
+      if !genesis_ended && session_changed {
+        Self::update_participants();
```

Comment:
Why can't we read the participants from validator-sets?

Reply:
We can and we already do within `update_participants`.  It is just that we keep the participants of the last session in this pallet since rewards are distributed to them in the current session. And we update them per session.

Reply:
I think validator-sets, in general, needs to have its database handling improved. There's a variety of specific behavior it's optimized around, yet as we've evolved it and extended it, it's drifted away from that behavior and lead to... wherever we are now.

validator-sets, upon its database handling cleanup/redo, should be able to service this functionality. emissions should not *need* to do this. If we need to accept that for now, fine, but we should consider and mark that.